pkg/resources/poddistribution/poddistribution.go

```diff
@@ -0,0 +1,386 @@
+package poddistribution
+
+import (
+	"context"
+	"fmt"
+	integreatlyv1alpha1 "github.com/integr8ly/integreatly-operator/pkg/apis/integreatly/v1alpha1"
+	appsv1 "github.com/openshift/api/apps/v1"
+	"github.com/sirupsen/logrus"
+	k8appsv1 "k8s.io/api/apps/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8sTypes "k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+	k8sclient "sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"strconv"
+	"strings"
+	"time"
+)
+
+const (
+	// ZoneLabel is the label that specifies the zone where a node is
+	ZoneLabel = "topology.kubernetes.io/zone"
+	// Annotation counter on the pods controller, dc, rs, ss
+	PodBalanceAttempts = "pod-balance-attempts"
+	maxBalanceAttempts = 3
+	name               = "name"
+	ns                 = "ns"
+	kind               = "kind"
+)
+
+type KindNameSpaceName struct {
+	*k8sTypes.NamespacedName
+	Kind string
+}
+
+func (knn KindNameSpaceName) String() string {
+	return fmt.Sprintf("%s/%s/%s", knn.Kind, knn.Namespace, knn.Name)
+}
+
+// Check the PodBalanceAttempts, ensure less than maxBalanceAttempts
+func verifyBalanceCount(ctx context.Context, client k8sclient.Client, obj metav1.Object) bool {
+	ant := obj.GetAnnotations()
+	// If there are no annotations then 0 balance attempts have been made
+	if ant == nil {
+		return true
+	}
+	if val, ok := ant[PodBalanceAttempts]; ok {
+		i, err := strconv.Atoi(val)
+		if err != nil {
+			logrus.Errorf("Error converting string annotations %s", ant[PodBalanceAttempts])
+			return false
+		} else {
+			if i >= maxBalanceAttempts {
+				logrus.Warningf("Reached max balance attempts for %s on %s", obj.GetName(), obj.GetNamespace())
+				return false
+			}
+		}
+	}
+
+	return true
+}
+
+func getObject(ctx context.Context, client k8sclient.Client, objDtls map[string]string) (metav1.Object, error) {
+	if objDtls[kind] == "DeploymentConfig" {
+		return getDeploymentConfig(ctx, client, objDtls)
+	}
+	if objDtls[kind] == "ReplicaSet" {
+		return getReplicaSet(ctx, client, objDtls)
+	}
+	if objDtls[kind] == "StatefulSet" {
+		return getStatefulSet(ctx, client, objDtls)
+	}
+	logrus.Errorf("Unexpected obj kind %s", objDtls[kind])
+	return nil, nil
+}
+
+func getDeploymentConfig(ctx context.Context, client k8sclient.Client, objDtls map[string]string) (metav1.Object, error) {
+	deploymentConfig := &appsv1.DeploymentConfig{}
+	err := client.Get(ctx, k8sTypes.NamespacedName{Name: objDtls[name], Namespace: objDtls[ns]}, deploymentConfig)
+	if err != nil {
+		logrus.Errorf("Error getting DeploymentConfig %v", err)
```

Comment:
does it makes sense to log here? Or better to return err, deployment config and let the caller figure out what to do with the error?

Reply:
done